### PR TITLE
Video Store _ Christina & Maryam

### PR DESCRIPTION
**Pull request by marshi23**
# Video Store API
Congratulations! You're submitting your assignment!
If you didn't get to the functionality the question is asking about, reply with what you would have done if you had completed it.

## Comprehension Questions
Question | Answer
:------------- | :-------------
Explain how you came up with the design of your ERD, based on the seed data. | The seed data was helpful to know what data we needed to input to our tables and columns.
What would be the Big-O time complexity of your /customers & /movies endpoints? What does the time complexity depend on? Explain your reasoning. | O(n) because it depends on the length of customers or movies in the db.
What is the Big-O time complexity of the `POST /rentals/check-in` endpoint? What does the time complexity depend on? Explain your reasoning.  | Constant because we're returning a JSON response that it was successful or unsuccessful.
Describe a set of positive and negative test cases you implemented for a model. | we tested for the validity of a movie given the required fields and then tested it without the required fields 
Describe a set of positive and negative test cases you implemented for a controller. | In our rental controller we test that the check-out path works with a valid input params and one without. 
How does your API respond when bad data is sent to it? | It responds with a 404 bad request status
Describe one of your custom model methods and why you chose to wrap that functionality into a method. | We have a checkout! class method in our rental model that takes in movie and customer ids. It adjusts and creates a rental object, makes note of the date, assigns a return date, updates the movie's inventory and the customer's movies_checked_out count. 
Do you have any recommendations on how we could improve this project for the next cohort? | Do this before API muncher, explain using postman better in relations to creating custom APIs 
Link to Trello |  https://trello.com/b/RZ6m8bHB
Link to ERD | https://www.lucidchart.com/invitations/accept/7b052de5-82d0-49e6-bb27-bba6572e18f8
Summary | We feel like we learned alot on this project, especially because we had enough project time to understand the concepts.

### CODE REVIEW

File: app/controllers/movies_controller.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+    else
+
+      render json:  movie.as_json(only: [:title, :overview, :release_date, :inventory]), status: :ok
+    end
+  end
+
+    def create
+      @movie = Movie.new(movie_params)
+      @movie.save!
```
Comment (CheezItMan) [inline app/controllers/movies_controller.rb:22]: You should return a `json` response if the move wasn't able to be saved.  Provide error messages.

File: app/controllers/rentals_controller.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
+class RentalsController < ApplicationController
+
+  def checkout
+
+    if rental_params[:customer_id] && rental_params[:movie_id]
+      rental = Rental.checkout!(rental_params[:customer_id], rental_params[:movie_id])
+      render json: { ok: true, message: 'Checkout successful!' }, status: :ok
+    else
+      render json: { ok: false, message: 'Unable to checkout' },
```
Comment (CheezItMan) [inline app/controllers/rentals_controller.rb:9]: This wouldn't be a not found but rather a bad request.

File: app/controllers/rentals_controller.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      status: :not_found
+    end
+
+  rescue ArgumentError
+    render json: { ok: false, message: 'Unable to checkout: movie or customer does not exist' }, status: :not_found
```
Comment (CheezItMan) [inline app/controllers/rentals_controller.rb:14]: Again, `not_found` is probably not the right response here.

File: test/models/movie_test.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      expect(movie.valid?).must_equal false
+    end
+  end
+
+  describe "relationships" do
+    let (:movie) { movies (:one) }
+
+    it "has rentals" do
+
+       movie.rentals.new
```
Comment (CheezItMan) [inline test/models/movie_test.rb:48]: What is this line trying to accomplish?

File: test/models/movie_test.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+      expect (movie.rentals.length).must_be :>=, 1
```
Comment (CheezItMan) [inline test/models/movie_test.rb:50]: I would also add a test that `rentals` is a collection of `Rental` instances.

File: test/controllers/movies_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+      {
+          title: 'Eternal sunshine of the Spotless Mind',
+          overview: 'Kate Winslet and Jim Carrey',
+          release_date: '2012-11-25',
+          inventory: "3"
+        }
+
+    end
+
+    it 'creates a movie given correct params' do
```
Comment (CheezItMan) [inline test/controllers/movies_controller_test.rb:65]: negative cases?

File: test/controllers/rentals_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+    it "adjusts inventory when a movie is checked out" do
+      rental = rentals(:one)
+      rental_params = {
+        customer_id: customers(:one)[:id],
+        movie_id: movies(:one)[:id]
+      }
+
+      post rentals_checkout_path(rental.id), {:params => rental_params}
+      rental.reload
```
Comment (CheezItMan) [inline test/controllers/rentals_controller_test.rb:90]: You also need negative cases as well.

**Comment by CheezItMan** [comment]
# Video Store
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** |
Git hygiene |  Good number of commits and good commit messages
Comprehension questions |  Check, I would say O(n) for checkin and check-out because you have to find the customer and movie.  You also should not return 404, for invalid data.  Instead a bad request would make an appropriate response in some cases.  
**General** |
Business Logic in Models |  Check, but you could have more in the customer and movie models.
All required endpoints return expected JSON data |  Check
Requests respond with appropriate HTTP Status Code |  Check
Errors are reported |  Not always, see my inline notes
**Testing** |
Passes all Smoke Tests |  You're not passing tests revolving around available inventory, and the movies checked out.  
Model Tests - all relations, validations, and custom functions test positive & negative cases |  You do not have tests for your relationships or custom methods in `Rental`.  Your testing is pretty thin here as well.
Controller Tests - URI parameters and data in the request body have positive & negative cases | You need to test negative cases here.  Like creating a movie with invalid parameters.  
**Overall** |  You hit **most** of the learning goals here.  You need to do a better job of testing throughly, adding more business logic in the models, and better matching the smoke tests.
